Source/JavaScriptCore/dfg/DFGInPlaceAbstractState.cpp

```diff
@@ -421,14 +418,289 @@ inline bool InPlaceAbstractState::mergeToSuccessors(BasicBlock* basicBlock)
+                falseNodes.append({ edge.node(), constant(edge.node(), jsNumber(0)) });
+                break;
+            case DoubleRepUse: {
+                AbstractValue trueValue = forNode(edge.node());
+                // We know the input is not NaN if it has a truthy value.
+                if (trueValue.filter(~SpecDoubleNaN) == FiltrationOK)
+                    trueNodes.append({ edge.node(), trueValue });
+                break;
+            }
+            case StringUse:
+                falseNodes.append({ edge.node(), constant(edge.node(), m_graph.m_vm.smallStrings.emptyString()) });
+                break;
+            case ObjectOrOtherUse: {
+                if (!m_graph.isWatchingMasqueradesAsUndefinedWatchpointSet(edge.node()))
+                    break;
+                AbstractValue trueValue = forNode(edge);
+                AbstractValue falseValue = trueValue;
+                if (trueValue.filter(SpecObject) == FiltrationOK)
+                    trueNodes.append({ edge.node(), trueValue });
+                if (falseValue.filter(SpecOther) == FiltrationOK)
+                    falseNodes.append({ edge.node(), falseValue });
+                break;
+            }
+            case BooleanUse:
+            case KnownBooleanUse:
+                falseNodes.append({ edge.node(), constant(edge.node(), jsBoolean(false)) });
+                trueNodes.append({ edge.node(), constant(edge.node(), jsBoolean(true)) });
+
+                switch (edge->op()) {
+                case CompareEq:
+                case CompareStrictEq: {
+                    Node* compare = edge.node();
+
+                    bool isValidCompare = [&]() {
+                        if (compare->op() == CompareStrictEq)
+                            return true;
+                        return compare->isBinaryUseKind(Int32Use)
+#if USE(JSVALUE64)
+                            || compare->isBinaryUseKind(Int52RepUse)
+#endif
+                            || compare->isBinaryUseKind(DoubleRepUse)
+                            || compare->isBinaryUseKind(StringUse)
+                            || compare->isBinaryUseKind(StringIdentUse)
+                            || compare->isBinaryUseKind(BooleanUse)
+                            || compare->isBinaryUseKind(ObjectUse)
+                            || compare->isBinaryUseKind(SymbolUse);
+                    }();
+                    if (!isValidCompare)
+                        break;
+
+                    AbstractValue lhs = forNode(compare->child1());
+                    AbstractValue rhs = forNode(compare->child2());
+                    if ((lhs.m_type & SpecDoubleNaN) || (rhs.m_type & SpecDoubleNaN)) {
+                        // We can't say anything useful about this comparison if either
+                        // input could be NaN, since it doesn't tell us any information
+                        // about the underlying values.
+                        break;
+                    }
```

Comment:
If either input is `NaN`, then the result must be false right. So, is that info not useful to the optimizer?
```
console.log(NaN == NaN);   // false
console.log(NaN === NaN);  // false
console.log(NaN == 5);      // false
console.log(NaN === 5);      // false
```

Reply:
`& SpecDoubleNaN` just says that the value *can* include NaN. So we cannot use this into false since this *can* be non-NaN.